validity/src/proposer.rs

```diff
@@ -188,74 +199,140 @@ where
                 return Ok(());
             }
         };
+    
+        let gas_threshold = self.program_config.gas_threshold;
+        let txs_threshold = self.program_config.txs_threshold;
+        let blocks_threshold = self.program_config.blocks_threshold;
+        let mut new_range_requests = Vec::new();
+    
+        // Determine request mode (Real or Mock)
+        let mode = if self.requester_config.mock {
+            RequestMode::Mock
+        } else {
+            RequestMode::Real
+        };
+    
+        let l1_chain_id = self.requester_config.l1_chain_id;
+        let l2_chain_id = self.requester_config.l2_chain_id;
+    
+        if gas_threshold > 0 || txs_threshold > 0 || blocks_threshold > 0 {
+            info!(
+                "Evaluating from block {} to block {} (gas threshold = {}, txs threshold = {}, blocks threshold = {})",
+                latest_proposed_block_number, finalized_block_number, gas_threshold, txs_threshold, blocks_threshold
+            );
+    
+            let mut start_block = self
+                .driver_config
+                .driver_db_client
+                .get_max_end_block(
+                    &self.program_config.commitments,
+                    l1_chain_id,
+                    l2_chain_id,
+                )
+                .await?;
+    
+            if start_block == 0 {
+                start_block = latest_proposed_block_number as i64;
+            }
```

Comment:
Reiterating on this, I don't think we need this, instead we should follow the same workflow as in the preset ranges:

1. Take the `latest_proposed_block_number`
2. Then `fetch_ranges_after_block`
3. Then `find_gaps` is needed
4. Then you can continue with the `create_range_requests_respecting_thresholds` or the `get_ranges_to_prove` + `create_range_request`

This way more code between the two code paths is maintained.

That being said, you don't need to create the new `get_max_end_block` as there's already an `fetch_highest_end_block_for_range_request`.